### PR TITLE
 arm64: dts: qcom: msm8916-vivo-y21l: Add initial device tree

### PR DESCRIPTION
**Pull request by raihan2000**
- [x] USB Networking
- [x] Flashing
- [x] WiFi
- [x] Bluetooth
- [x] Display
- [ ] Audio
- [ ] Battery
- [x] Accelerometer
- [x] Ambient Light
- [x] Magnetometer

Notes: touchscreen uses synaptics_rmi driver but it does not load at the momment, need to modify it from downstream, display works but drm is broken, smb358-charger is in driver need to add

### CODE REVIEW

File: arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	status = "okay";
+};
+
+&usb {
+	status = "okay";
+};
+
+&usb {
+    extcon = <&usb_id>, <&usb_id>;
+    status = "okay";
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts]:
```suggestion
	extcon = <&usb_id>, <&usb_id>;
	status = "okay";
```

Use tabs for indents.

File: arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+};
+
+&usb_hs_phy {
+    extcon = <&usb_id>;
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts]:
```suggestion
	extcon = <&usb_id>;
```

File: arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	status = "okay";
+};
+
+&sdhc_1 {
+	status = "okay";
+};
+
+&usb {
+	status = "okay";
+};
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts]: Duplicated?

**Comment by wonderfulShrineMaidenOfParadise** [comment]
Does it support SD cards? Perhaps you can add it to the initial commit as well.

File: arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+			panel_in: endpoint {
+				remote-endpoint = <&mdss_dsi0_out>;
+			};
+		};
+	};
+};
+
+&mdss_dsi0_out {
+    data-lanes = <0 1>;
+    remote-endpoint = <&panel_in>;
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts]:
```suggestion
	data-lanes = <0 1>;
	remote-endpoint = <&panel_in>;
```

File: arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+};
+
+&pm8916_mpps {
+    pwm_out: mpp4-state {
+        pins = "mpp4";
+        function = "digital";
+        power-source = <PM8916_MPP_VPH>;
+        output-low;
+        qcom,dtest = <1>;
+    };
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts]:
```suggestion
	pwm_out: mpp4-state {
		pins = "mpp4";
		function = "digital";
		power-source = <PM8916_MPP_VPH>;
		output-low;
		qcom,dtest = <1>;
	};
```

pinctrls are preferred to be put at the bottom of a device tree. `&pm8916_mpps` is usually after `&tlmm`.

File: arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+&sdhc_1 {
+	vmmc-supply = <&pm8916_l8>;
+	vqmmc-supply = <&pm8916_l5>;
+
+	status = "okay";
+};
+
+&sdhc_2 {
+	status = "okay";
+	non-removable;
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts]:
```suggestion
	non-removable;
	status = "okay";
```

Always put `status` at the bottom.

File: arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+        pins = "mpp4";
+        function = "digital";
+        power-source = <PM8916_MPP_VPH>;
+        output-low;
+        qcom,dtest = <1>;
+    };
+};
+
+&pm8916_pwm {
+	status = "okay";
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts:81]: ^

File: drivers/gpu/drm/panel/msm8916-generated/Makefile [hunk 8]
```diff
@@ -19,6 +19,9 @@ obj-$(CONFIG_DRM_PANEL_MOTOROLA_HARPIA_TIANMA) += panel-motorola-harpia-tianma.o
 obj-$(CONFIG_DRM_PANEL_MOTOROLA_OSPREY_INX) += panel-motorola-osprey-inx.o
 obj-$(CONFIG_DRM_PANEL_MOTOROLA_SURNIA_BOE) += panel-motorola-surnia-boe.o
 obj-$(CONFIG_DRM_PANEL_MOTOROLA_SURNIA_INX) += panel-motorola-surnia-inx.o
+obj-$(CONFIG_DRM_PANEL_NT35510S_TM) += panel-nt35510s-tm.o
+obj-$(CONFIG_DRM_PANEL_ORISE8012A_TMFWVGA) += panel-orise8012a-tmfwvga.o
+obj-$(CONFIG_DRM_PANEL_ORISE8012A_TMFWVGA_BKG) += panel-orise8012a-tmfwvga-bkg.o
```
Comment (wonderfulShrineMaidenOfParadise) [inline drivers/gpu/drm/panel/msm8916-generated/Makefile]: Same here. Keep only what you have and drop the rest.

File: arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+};
+
+&mdss_dsi0 {
+	pinctrl-names = "default", "sleep";
+	pinctrl-0 = <&mdss_default>;
+	pinctrl-1 = <&mdss_sleep>;
+
+	panel@0 {
+		compatible ="vivo,orise8012a-tmfwvga", "mdss,orise8012a-tmfwvga";
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts]:
```suggestion
		compatible = "vivo,y21l-panel";
```

This is what we added in lk2nd, for panel selection.

File: drivers/gpu/drm/panel/msm8916-generated/panel-orise8012a-tmfwvga.c [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+
+	ret = mipi_dsi_detach(dsi);
+	if (ret < 0)
+		dev_err(&dsi->dev, "Failed to detach from DSI host: %d\n", ret);
+
+	drm_panel_remove(&ctx->panel);
+}
+
+static const struct of_device_id orise8012a_tmfwvga_of_match[] = {
+	{ .compatible = "mdss,orise8012a-tmfwvga" }, // FIXME
```
Comment (wonderfulShrineMaidenOfParadise) [inline drivers/gpu/drm/panel/msm8916-generated/panel-orise8012a-tmfwvga.c]: Use https://github.com/msm8916-mainline/linux-panel-drivers scripts to generate the driver. You need to add a new config and dtb to the repo.

File: arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+		brightness-levels = <0 255>;
+		num-interpolated-steps = <255>;
+		default-brightness-level = <255>;
+	};
+
+	gpio-keys {
+		compatible = "gpio-keys";
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&gpio_keys_default>;
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts]:
```suggestion
		pinctrl-0 = <&gpio_keys_default>;
		pinctrl-names = "default";
```

Not my idea, but suggested by upstream.

File: arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+			gpios = <&tlmm 107 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_VOLUMEUP>;
+		};
+	};
+
+	usb_id: usb-id {
+		compatible = "linux,extcon-usb-gpio";
+		id-gpios = <&tlmm 110 GPIO_ACTIVE_HIGH>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&usb_id_default>;
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts]:
```suggestion
		pinctrl-0 = <&usb_id_default>;
		pinctrl-names = "default";
```

File: arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+};
+
+&mdss {
+	status = "okay";
+};
+
+&mdss_dsi0 {
+	pinctrl-names = "default", "sleep";
+	pinctrl-0 = <&mdss_default>;
+	pinctrl-1 = <&mdss_sleep>;
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts]:
```suggestion
	pinctrl-0 = <&mdss_default>;
	pinctrl-1 = <&mdss_sleep>;
	pinctrl-names = "default", "sleep";
```

File: arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+        power-source = <PM8916_MPP_VPH>;
+        output-low;
+        qcom,dtest = <1>;
+    };
+};
+
+&pm8916_pwm {
+	status = "okay";
+	pinctrl-names = "default";
+	pinctrl-0 = <&pwm_out>;
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts]:
```suggestion
	pinctrl-0 = <&pwm_out>;
	pinctrl-names = "default";
```

File: arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+
+&wcnss_mem {
+	status = "okay";
+};
+
+&tlmm {
+	gpio_keys_default: gpio-keys-default-state {
+		pins = "gpio107";
+		function = "gpio";
+
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts]:
```suggestion
```

Suggested by upstream as well.

File: arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+		drive-strength = <2>;
+		bias-pull-up;
+	};
+
+	mdss_default: mdss-default-state {
+		pins = "gpio25";
+		function = "gpio";
+
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts]:
```suggestion
```

File: arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+		drive-strength = <8>;
+		bias-disable;
+	};
+
+	mdss_sleep: mdss-sleep-state {
+		pins = "gpio25";
+		function = "gpio";
+
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts]:
```suggestion
```

File: arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+		drive-strength = <2>;
+		bias-pull-down;
+	};
+
+	usb_id_default: usb-id-default-state {
+		pins = "gpio110";
+		function = "gpio";
+
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts]:
```suggestion
```

File: drivers/gpu/drm/panel/msm8916-generated/panel-orise8012a-tmfwvga.c (hunk 10 above)
Comment (raihan2000) [reply drivers/gpu/drm/panel/msm8916-generated/panel-orise8012a-tmfwvga.c]: is there need to add downstream dtb??

File: arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts [hunk 19]
```diff
@@ -0,0 +1,234 @@
...
+&mdss {
+	status = "okay";
+};
+
+&mdss_dsi0 {
+	pinctrl-0 = <&mdss_default>;
+	pinctrl-1 = <&mdss_sleep>;
+	pinctrl-names = "default", "sleep";
+	panel@0 {
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts]:
```suggestion
	pinctrl-names = "default", "sleep";

	panel@0 {
```
Always have a newline between properties and nodes.

File: arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts [hunk 20]
```diff
@@ -0,0 +1,234 @@
...
+		compatible ="vivo,y21l-panel";
+		reg = <0>;
+		backlight = <&backlight>;
+		reset-gpios = <&tlmm 25 GPIO_ACTIVE_LOW>;
+		power-supply = <&pm8916_l17>;
+		port {
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts]:
```suggestion
		power-supply = <&pm8916_l17>;

		port {
```

File: arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts [hunk 21]
```diff
@@ -0,0 +1,234 @@
...
+		num-interpolated-steps = <255>;
+		default-brightness-level = <255>;
+	};
+
+	gpio-keys {
+		compatible = "gpio-keys";
+		pinctrl-0 = <&gpio_keys_default>;
+		pinctrl-names = "default";
+		label = "GPIO Buttons";
+		button-volume-up {
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts]:
```suggestion
		label = "GPIO Buttons";

		button-volume-up {
```

File: drivers/gpu/drm/panel/msm8916-generated/panel-orise8012a-tmfwvga.c (hunk 10 above)
Comment (wonderfulShrineMaidenOfParadise) [reply drivers/gpu/drm/panel/msm8916-generated/panel-orise8012a-tmfwvga.c]: Yes, downstream dtb extracted from boot/recovery.img

File: arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts [hunk 22]
```diff
@@ -0,0 +1,234 @@
...
+				"0", "0", "1";
+		pinctrl-0 = <&accel_int_default>;
+		pinctrl-names = "default";
+	};
+
+	light-sensor@39 {
+		compatible = "amstaos,tmd2772";
+		reg = <0x39>;
+		interrupt-parent = <&tlmm>;
+		interrupts = <113 IRQ_TYPE_EDGE_FALLING>;
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts]:
```suggestion
		interrupts-extended = <&tlmm 113 IRQ_TYPE_EDGE_FALLING>;
```

File: arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts [hunk 23]
```diff
@@ -0,0 +1,234 @@
...
+};
+
+&blsp_i2c2 {
+	status = "okay";
+
+	accelerometer@18 {
+		compatible = "st,lis3dh-accel";
+		reg = <0x18>;
+		interrupt-parent = <&tlmm>;
+		interrupts = <21 IRQ_TYPE_EDGE_RISING>;
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts]:
```suggestion
		interrupts-extended = <&tlmm 21 IRQ_TYPE_EDGE_RISING>;
```

Just leave a note here. It seems to be a weak preference in upstream.

File: drivers/gpu/drm/panel/msm8916-generated/panel-orise8012a-tmfwvga.c (hunk 10 above)
Comment (raihan2000) [reply drivers/gpu/drm/panel/msm8916-generated/panel-orise8012a-tmfwvga.c]: the device is there https://github.com/msm8916-mainline/linux-panel-drivers/pull/33

File: arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts [hunk 24]
```diff
@@ -0,0 +1,234 @@
...
+	status = "okay";
+};
+
+&mdss_dsi0 {
+	pinctrl-0 = <&mdss_default>;
+	pinctrl-1 = <&mdss_sleep>;
+	pinctrl-names = "default", "sleep";
+
+	panel@0 {
+		compatible ="vivo,y21l-panel";
```
Comment (wonderfulShrineMaidenOfParadise) [inline arch/arm64/boot/dts/qcom/msm8916-vivo-y21l.dts]:
```suggestion
		compatible = "vivo,y21l-panel";
```

**Comment by stephan-gh** [comment]
Please squash your commits and do cleanly separated ones with a proper commit message, so that  "check (pull request)" passes. Feel free to ask in the chat if you need help with Git.

**Pull request closed without merging**